src/System.Private.CoreLib/shared/System/Text/Unicode/Utf8Utility.cs

```diff
@@ -44,6 +60,143 @@ public static unsafe int GetIndexOfFirstInvalidUtf8Sequence(ReadOnlySpan<byte> u
         }
 
 #if FEATURE_UTF8STRING
+        /// <summary>
+        /// Given a buffer <paramref name="utf8Data"/> which represents guaranteed well-formed UTF-8 data,
+        /// returns the number of <see langword="char"/>s which would result from transcoding the buffer
+        /// to UTF-16. The behavior of this method is undefined if <paramref name="utf8Data"/> contains
+        /// any ill-formed UTF-8 subsequences.
+        /// </summary>
+        public static unsafe int GetUtf16CharCountFromKnownWellFormedUtf8(ReadOnlySpan<byte> utf8Data)
+        {
+            // Remember: the number of resulting UTF-16 chars will never be greater than the number
+            // of UTF-8 bytes given well-formed input, so we can get away with casting the final
+            // result to an 'int'.
+
+            fixed (byte* pPinnedUtf8Data = &MemoryMarshal.GetReference(utf8Data))
+            {
+                if (Sse2.IsSupported && Popcnt.IsSupported)
```

Comment:
```suggestion
                if (Popcnt.IsSupported)
```
`Popcnt` will only be supported iif `Sse2` is supported.
The JIT will branch-eliminate anyway, so no problem to keep it more verbose here.

`Sse2` is only available on little-endian, but you have an assert below. So you could add an assert for `Sse2` and simplify the condition.
Personally I would omit the asserts for little-endian and for sse2. If there's something wrong, then it's a (big) JIT-bug.

Reply:
> Sse2 is only available on little-endian

Sse2 is not available on ARM that is also little-endian.

Reply:
Let's rephrase my intention: `Sse2.IsSupported` implies little-endian, so the assert could be omitted.

Reply:
In general I prefer to keep the additional assertions, because it's a further stake in the ground that says "BTW, though it may not be obvious from the code that follows, this is only designed and tested with little endian in mind."